File: re/ghidra/scripts/FindWagesArtifacts.py
# FindWagesArtifacts.py — Ghidra headless script
#
# Purpose:
#   First-pass investigation of Wow.exe (Wages of War, 1996, MSVC 4.x x86).
#   Exports observational data we use to ground-truth the clean-room engine:
#     - all defined strings with addresses
#     - all functions with entry points + signatures
#     - all DLL imports
#     - strings filtered to ones that look like data-file path patterns
#       (TIL, OBJ, DAT, MAP, PCX, WAV, MID, COR, VLS, VLA, WRI, .EXE, MAPS\, scen, mish, mssn)
#     - for each filtered string, the list of functions that reference it (xrefs)
#
#   This is OBSERVATIONAL — facts about offsets and references in the binary.
#   It is NOT vendor source. Output goes to whiteroom/analysis/ and never
#   into crates/.
#
# Output:
#   <project_dir>/../analysis/Wow.exe-strings.json
#   <project_dir>/../analysis/Wow.exe-functions.json
#   <project_dir>/../analysis/Wow.exe-imports.json
#   <project_dir>/../analysis/Wow.exe-data-file-xrefs.json
#   <project_dir>/../analysis/Wow.exe-summary.md
#
# Usage (headless):
#   analyzeHeadless.bat <project_dir> <project_name> -process Wow.exe -postScript FindWagesArtifacts.py
#
#@category wages-of-war

import json
import os
import re

program = getCurrentProgram()
if program is None:
    print("FindWagesArtifacts: no program loaded")
    exit(1)

prog_name = program.getName()
listing = program.getListing()
fm = program.getFunctionManager()
st = program.getSymbolTable()
mem = program.getMemory()
ref_mgr = program.getReferenceManager()

# ---------- Output paths ----------
project_location = program.getDomainFile().getParent().getProjectLocator().getProjectDir()
analysis_dir = os.path.abspath(os.path.join(str(project_location), "..", "analysis"))
if not os.path.isdir(analysis_dir):
    os.makedirs(analysis_dir)

def out(name):
    return os.path.join(analysis_dir, prog_name + "-" + name)

# ---------- Helper: address -> containing-function name ----------
def func_at(addr):
    f = fm.getFunctionContaining(addr)
    return f.getName() if f else "<no_func>"

# ---------- Strings ----------
print("Collecting strings...")
strings_all = []
data_iter = listing.getDefinedData(True)
for d in data_iter:
    dt = d.getDataType()
    if dt is None:
        continue
    nm = dt.getName().lower()
    # Catch C strings, Unicode strings, string-array data types
    if "string" in nm or "char[" in nm or nm == "char":
        try:
            v = d.getValue()
            if v is None:
                continue
            s = str(v)
            if len(s) < 3:
                continue
            strings_all.append({
                "addr": "0x%s" % d.getAddress().toString(),
                "addr_offset": d.getAddress().getOffset(),
                "type": dt.getName(),
                "len": d.getLength(),
                "value": s,
            })
        except Exception:
            continue

print("  %d defined strings" % len(strings_all))

# ---------- Functions ----------
print("Collecting functions...")
funcs_all = []
for func in fm.getFunctions(True):
    entry = func.getEntryPoint()
    body_size = 0
    try:
        body_size = func.getBody().getNumAddresses()
    except Exception:
        pass
    funcs_all.append({
        "name": func.getName(),
        "entry": "0x%s" % entry.toString(),
        "entry_offset": entry.getOffset(),
        "size_addrs": int(body_size),
        "is_thunk": bool(func.isThunk()),
        "is_external": bool(func.isExternal()),
        "calling_convention": (
            func.getCallingConventionName() if func.getCallingConventionName() else ""
        ),
        "signature": func.getSignature().toString(),
    })
print("  %d functions" % len(funcs_all))

# ---------- Imports (DLL functions) ----------
print("Collecting imports...")
imports_by_dll = {}
for ext in st.getExternalSymbols():
    src = ext.getParentNamespace()
    dll = src.getName() if src else "<unknown>"
    imports_by_dll.setdefault(dll, []).append({
        "name": ext.getName(),
        "addr": "0x%s" % ext.getAddress().toString() if ext.getAddress() else None,
    })
print("  %d DLLs, %d total imports" % (
    len(imports_by_dll),
    sum(len(v) for v in imports_by_dll.values()),
))

# ---------- Data-file xref strings ----------
# These patterns target known WoW data-file references the engine expects.
# Wow.exe should contain printf/fopen format strings or filenames that the
# loader uses to find each scenario's TIL/OBJ/DAT/MAP/etc.
patterns = [
    ("TIL", re.compile(r"\.TIL\b|TIL\d|TILSCN", re.I)),
    ("OBJ", re.compile(r"\.OBJ\b|OBJ\d", re.I)),
    ("MAP", re.compile(r"\.MAP\b|SCEN.*MAP|MAPS[\\/]", re.I)),
    ("DAT", re.compile(r"\.DAT\b|TILES\d|JUNGSLD|RIFLWALK", re.I)),
    ("PCX", re.compile(r"\.PCX\b|OFFPIC|SCENPIC", re.I)),
    ("PAL", re.compile(r"\.PAL\b|PALETTE", re.I)),
    ("WAV", re.compile(r"\.WAV\b", re.I)),
    ("MID", re.compile(r"\.MID\b", re.I)),
    ("COR", re.compile(r"\.COR\b", re.I)),
    ("VLS", re.compile(r"\.VLS\b|\.VLA\b", re.I)),
    ("WRI", re.compile(r"\.WRI\b|MISHN|MSSN", re.I)),
    ("BTN", re.compile(r"\.BTN\b|BUTTONS", re.I)),
    ("CONTRACT", re.compile(r"CONTRACT|HIRE|MERC", re.I)),
    ("FORMAT_PRINTF", re.compile(r"%[-+ 0#]?[0-9]*[hl]?[a-zA-Z]")),
]

xref_strings = []
for s in strings_all:
    val = s["value"]
    matched_cats = []
    for (cat, rx) in patterns:
        if rx.search(val):
            matched_cats.append(cat)
    if not matched_cats:
        continue
    addr = toAddr(s["addr_offset"])
    refs = []
    for r in ref_mgr.getReferencesTo(addr):
        from_addr = r.getFromAddress()
        refs.append({
            "from": "0x%s" % from_addr.toString(),
            "from_offset": from_addr.getOffset(),
            "from_func": func_at(from_addr),
            "ref_type": r.getReferenceType().getName(),
        })
    xref_strings.append({
        "addr": s["addr"],
        "addr_offset": s["addr_offset"],
        "value": val,
        "categories": matched_cats,
        "xref_count": len(refs),
        "xrefs": refs,
    })

# Sort by xref_count descending — the most-referenced data-file strings
# are the high-value investigation targets.
xref_strings.sort(key=lambda x: -x["xref_count"])
print("  %d data-file-related strings with xrefs" % len(xref_strings))

# ---------- Write JSON outputs ----------
def write_json(name, data):
    p = out(name + ".json")
    with open(p, "w") as f:
        json.dump(data, f, indent=2)
    print("  wrote %s" % p)

write_json("strings", strings_all)
write_json("functions", funcs_all)
write_json("imports", imports_by_dll)
write_json("data-file-xrefs", xref_strings)

# ---------- Markdown summary ----------
md_path = out("summary.md")
with open(md_path, "w") as f:
    f.write("# Wow.exe — first-pass artifact summary\n\n")
    f.write("Generated by `FindWagesArtifacts.py` against `%s`.\n\n" % prog_name)
    f.write("- **%d defined strings**\n" % len(strings_all))
    f.write("- **%d functions**\n" % len(funcs_all))
    f.write("- **%d external DLL imports across %d DLLs**\n" % (
        sum(len(v) for v in imports_by_dll.values()), len(imports_by_dll),
    ))
    f.write("- **%d data-file-related strings with xrefs**\n\n" % len(xref_strings))

    f.write("## DLL imports (sorted)\n\n")
    for dll in sorted(imports_by_dll.keys()):
        funcs = imports_by_dll[dll]
        f.write("### `%s` (%d funcs)\n\n" % (dll, len(funcs)))
        for fn in sorted(funcs, key=lambda x: x["name"]):
            f.write("- `%s`\n" % fn["name"])
        f.write("\n")

    f.write("## Top data-file-related strings (most-referenced first)\n\n")
    f.write("| String | Categories | Xref count | Calling functions |\n")
    f.write("|--------|------------|------------|-------------------|\n")
    for s in xref_strings[:80]:
        funcs = sorted(set(r["from_func"] for r in s["xrefs"]))
        f.write("| `%s` | %s | %d | %s |\n" % (
            s["value"].replace("|", "\\|").replace("\n", "\\n")[:80],
            ",".join(s["categories"]),
            s["xref_count"],
            ", ".join(funcs)[:120],
        ))

print("  wrote %s" % md_path)
print("FindWagesArtifacts: done")

File: re/ghidra/scripts/FindWagesArtifacts.java
// FindWagesArtifacts.java — Ghidra headless script (Java)
//
// First-pass investigation of Wow.exe (Wages of War, 1996, MSVC 4.x x86).
// Exports observational data we use to ground-truth the clean-room engine.
// OUTPUT IS OBSERVATIONAL FACTS ABOUT THE BINARY — never copy of vendor source.
// Output goes to <project_dir>/../analysis/ and never into crates/.
//
//@category wages-of-war

import ghidra.app.script.GhidraScript;
import ghidra.program.model.address.Address;
import ghidra.program.model.data.DataType;
import ghidra.program.model.listing.Data;
import ghidra.program.model.listing.Function;
import ghidra.program.model.listing.FunctionManager;
import ghidra.program.model.listing.Listing;
import ghidra.program.model.listing.Program;
import ghidra.program.model.symbol.ExternalLocation;
import ghidra.program.model.symbol.ExternalLocationIterator;
import ghidra.program.model.symbol.ExternalManager;
import ghidra.program.model.symbol.Reference;
import ghidra.program.model.symbol.ReferenceManager;

import java.io.File;
import java.io.FileWriter;
import java.io.IOException;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.Comparator;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.TreeMap;
import java.util.TreeSet;
import java.util.regex.Pattern;

public class FindWagesArtifacts extends GhidraScript {

    @Override
    public void run() throws Exception {
        Program program = currentProgram;
        if (program == null) {
            println("FindWagesArtifacts: no program loaded");
            return;
        }
        String progName = program.getName();
        Listing listing = program.getListing();
        FunctionManager fm = program.getFunctionManager();
        ReferenceManager refMgr = program.getReferenceManager();
        ExternalManager extMgr = program.getExternalManager();

        File projectDir = program.getDomainFile()
            .getParent().getProjectLocator().getProjectDir();
        File analysisDir = new File(projectDir.getParentFile(), "analysis");
        if (!analysisDir.isDirectory()) {
            analysisDir.mkdirs();
        }

        // ---------- Strings ----------
        println("Collecting strings...");
        List<Map<String,Object>> stringsAll = new ArrayList<>();
        var dataIter = listing.getDefinedData(true);
        while (dataIter.hasNext()) {
            Data d = dataIter.next();
            DataType dt = d.getDataType();
            if (dt == null) continue;
            String tname = dt.getName().toLowerCase();
            if (!(tname.contains("string") || tname.startsWith("char[") || tname.equals("char"))) continue;
            Object v = d.getValue();
            if (v == null) continue;
            String s = v.toString();
            if (s.length() < 3) continue;
            Map<String,Object> m = new HashMap<>();
            m.put("addr", "0x" + d.getAddress().toString());
            m.put("addr_offset", d.getAddress().getOffset());
            m.put("type", dt.getName());
            m.put("len", d.getLength());
            m.put("value", s);
            stringsAll.add(m);
        }
        println("  " + stringsAll.size() + " defined strings");

        // ---------- Functions ----------
        println("Collecting functions...");
        List<Map<String,Object>> funcsAll = new ArrayList<>();
        for (Function f : fm.getFunctions(true)) {
            Address entry = f.getEntryPoint();
            long bodySize = 0;
            try {
                bodySize = f.getBody().getNumAddresses();
            } catch (Exception ignored) {}
            Map<String,Object> m = new HashMap<>();
            m.put("name", f.getName());
            m.put("entry", "0x" + entry.toString());
            m.put("entry_offset", entry.getOffset());
            m.put("size_addrs", bodySize);
            m.put("is_thunk", f.isThunk());
            m.put("is_external", f.isExternal());
            String cc = f.getCallingConventionName();
            m.put("calling_convention", cc == null ? "" : cc);
            m.put("signature", f.getSignature().toString());
            funcsAll.add(m);
        }
        println("  " + funcsAll.size() + " functions");

        // ---------- Imports (DLL functions) ----------
        println("Collecting imports...");
        Map<String, List<Map<String,Object>>> importsByDll = new TreeMap<>();
        String[] libs = extMgr.getExternalLibraryNames();
        int totalImports = 0;
        for (String lib : libs) {
            ExternalLocationIterator it = extMgr.getExternalLocations(lib);
            List<Map<String,Object>> funcs = new ArrayList<>();
            while (it.hasNext()) {
                ExternalLocation loc = it.next();
                Map<String,Object> m = new HashMap<>();
                m.put("name", loc.getLabel());
                Address a = loc.getAddress();
                m.put("addr", a == null ? null : "0x" + a.toString());
                funcs.add(m);
                totalImports++;
            }
            funcs.sort(Comparator.comparing(x -> ((String)x.get("name")).toLowerCase()));
            importsByDll.put(lib, funcs);
        }
        println("  " + libs.length + " DLLs, " + totalImports + " total imports");

        // ---------- Data-file xref strings ----------
        Object[][] patternData = {
            {"TIL", "(?i)\\.TIL\\b|TIL\\d|TILSCN"},
            {"OBJ", "(?i)\\.OBJ\\b|OBJ\\d"},
            {"MAP", "(?i)\\.MAP\\b|SCEN.*MAP|MAPS[\\\\/]"},
            {"DAT", "(?i)\\.DAT\\b|TILES\\d|JUNGSLD|RIFLWALK"},
            {"PCX", "(?i)\\.PCX\\b|OFFPIC|SCENPIC"},
            {"PAL", "(?i)\\.PAL\\b|PALETTE"},
            {"WAV", "(?i)\\.WAV\\b"},
            {"MID", "(?i)\\.MID\\b"},
            {"COR", "(?i)\\.COR\\b"},
            {"VLS", "(?i)\\.VLS\\b|\\.VLA\\b"},
            {"WRI", "(?i)\\.WRI\\b|MISHN|MSSN"},
            {"BTN", "(?i)\\.BTN\\b|BUTTONS"},
            {"CONTRACT", "(?i)CONTRACT|HIRE|MERC"},
            {"FORMAT_PRINTF", "%[-+ 0#]?[0-9]*[hl]?[a-zA-Z]"},
        };
        List<String[]> patterns = new ArrayList<>();
        List<Pattern> compiledPatterns = new ArrayList<>();
        for (Object[] row : patternData) {
            patterns.add(new String[]{(String)row[0], (String)row[1]});
            compiledPatterns.add(Pattern.compile((String)row[1]));
        }

        List<Map<String,Object>> xrefStrings = new ArrayList<>();
        for (Map<String,Object> s : stringsAll) {
            String val = (String) s.get("value");
            List<String> matchedCats = new ArrayList<>();
            for (int i = 0; i < patterns.size(); i++) {
                if (compiledPatterns.get(i).matcher(val).find()) {
                    matchedCats.add(patterns.get(i)[0]);
                }
            }
            if (matchedCats.isEmpty()) continue;
            long off = (long) s.get("addr_offset");
            Address addr = toAddr(off);
            List<Map<String,Object>> refs = new ArrayList<>();
            for (Reference r : refMgr.getReferencesTo(addr)) {
                Address fromAddr = r.getFromAddress();
                Function fn = fm.getFunctionContaining(fromAddr);
                Map<String,Object> rm = new HashMap<>();
                rm.put("from", "0x" + fromAddr.toString());
                rm.put("from_offset", fromAddr.getOffset());
                rm.put("from_func", fn == null ? "<no_func>" : fn.getName());
                rm.put("ref_type", r.getReferenceType().getName());
                refs.add(rm);
            }
            Map<String,Object> m = new HashMap<>();
            m.put("addr", s.get("addr"));
            m.put("addr_offset", off);
            m.put("value", val);
            m.put("categories", matchedCats);
            m.put("xref_count", refs.size());
            m.put("xrefs", refs);
            xrefStrings.add(m);
        }
        xrefStrings.sort((a, b) -> Integer.compare((int) b.get("xref_count"), (int) a.get("xref_count")));
        println("  " + xrefStrings.size() + " data-file-related strings with xrefs");

        // ---------- Write JSON outputs (hand-written; no Gson dep) ----------
        writeJson(new File(analysisDir, progName + "-strings.json"), stringsAll);
        writeJson(new File(analysisDir, progName + "-functions.json"), funcsAll);
        writeJson(new File(analysisDir, progName + "-imports.json"), importsByDll);
        writeJson(new File(analysisDir, progName + "-data-file-xrefs.json"), xrefStrings);

        // ---------- Markdown summary ----------
        File mdPath = new File(analysisDir, progName + "-summary.md");
        try (FileWriter fw = new FileWriter(mdPath)) {
            fw.write("# Wow.exe — first-pass artifact summary\n\n");
            fw.write("Generated by `FindWagesArtifacts.java` against `" + progName + "`.\n\n");
            fw.write("- **" + stringsAll.size() + " defined strings**\n");
            fw.write("- **" + funcsAll.size() + " functions**\n");
            fw.write("- **" + totalImports + " external DLL imports across " + libs.length + " DLLs**\n");
            fw.write("- **" + xrefStrings.size() + " data-file-related strings with xrefs**\n\n");

            fw.write("## DLL imports (sorted)\n\n");
            for (Map.Entry<String, List<Map<String,Object>>> e : importsByDll.entrySet()) {
                fw.write("### `" + e.getKey() + "` (" + e.getValue().size() + " funcs)\n\n");
                for (Map<String,Object> fn : e.getValue()) {
                    fw.write("- `" + fn.get("name") + "`\n");
                }
                fw.write("\n");
            }

            fw.write("## Top data-file-related strings (most-referenced first)\n\n");
            fw.write("| String | Categories | Xref count | Calling functions |\n");
            fw.write("|--------|------------|------------|-------------------|\n");
            int limit = Math.min(80, xrefStrings.size());
            for (int i = 0; i < limit; i++) {
                Map<String,Object> s = xrefStrings.get(i);
                @SuppressWarnings("unchecked")
                List<Map<String,Object>> refs = (List<Map<String,Object>>) s.get("xrefs");
                Set<String> funcSet = new TreeSet<>();
                for (Map<String,Object> r : refs) funcSet.add((String) r.get("from_func"));
                String val = ((String) s.get("value")).replace("|", "\\|").replace("\n", "\\n");
                if (val.length() > 80) val = val.substring(0, 80);
                @SuppressWarnings("unchecked")
                List<String> cats = (List<String>) s.get("categories");
                String funcsStr = String.join(", ", funcSet);
                if (funcsStr.length() > 120) funcsStr = funcsStr.substring(0, 120);
                fw.write("| `" + val + "` | " + String.join(",", cats)
                    + " | " + s.get("xref_count") + " | " + funcsStr + " |\n");
            }
        }
        println("  wrote " + mdPath.getAbsolutePath());
        println("FindWagesArtifacts: done");
    }

    // ----- naive JSON writer (good enough for our shapes; no quoting of keys
    // beyond replace; no nested escapes needed beyond \\ and \") -----
    private void writeJson(File path, Object obj) throws IOException {
        try (FileWriter fw = new FileWriter(path)) {
            fw.write(jsonOf(obj, 0));
            fw.write("\n");
        }
        println("  wrote " + path.getAbsolutePath());
    }

    private String jsonOf(Object o, int indent) {
        if (o == null) return "null";
        if (o instanceof Boolean) return o.toString();
        if (o instanceof Number) return o.toString();
        if (o instanceof String) return jsonString((String) o);
        if (o instanceof List) {
            List<?> list = (List<?>) o;
            if (list.isEmpty()) return "[]";
            StringBuilder sb = new StringBuilder();
            String pad = "  ".repeat(indent + 1);
            String padClose = "  ".repeat(indent);
            sb.append("[\n");
            for (int i = 0; i < list.size(); i++) {
                sb.append(pad).append(jsonOf(list.get(i), indent + 1));
                if (i < list.size() - 1) sb.append(",");
                sb.append("\n");
            }
            sb.append(padClose).append("]");
            return sb.toString();
        }
        if (o instanceof Map) {
            Map<?,?> map = (Map<?,?>) o;
            if (map.isEmpty()) return "{}";
            StringBuilder sb = new StringBuilder();
            String pad = "  ".repeat(indent + 1);
            String padClose = "  ".repeat(indent);
            sb.append("{\n");
            int i = 0;
            int n = map.size();
            for (Map.Entry<?,?> e : map.entrySet()) {
                sb.append(pad)
                  .append(jsonString(e.getKey().toString()))
                  .append(": ")
                  .append(jsonOf(e.getValue(), indent + 1));
                if (i < n - 1) sb.append(",");
                sb.append("\n");
                i++;
            }
            sb.append(padClose).append("}");
            return sb.toString();
        }
        return jsonString(o.toString());
    }

    private String jsonString(String s) {
        StringBuilder sb = new StringBuilder(s.length() + 2);
        sb.append('"');
        for (int i = 0; i < s.length(); i++) {
            char c = s.charAt(i);
            switch (c) {
                case '\\': sb.append("\\\\"); break;
                case '"':  sb.append("\\\""); break;
                case '\n': sb.append("\\n"); break;
                case '\r': sb.append("\\r"); break;
                case '\t': sb.append("\\t"); break;
                case '\b': sb.append("\\b"); break;
                case '\f': sb.append("\\f"); break;
                default:
                    if (c < 0x20) sb.append(String.format("\\u%04x", (int) c));
                    else sb.append(c);
            }
        }
        sb.append('"');
        return sb.toString();
    }
}
